File: lambda/javascript/upload-podcast.js
import { Upload } from '@aws-sdk/lib-storage';
import { request } from 'https';
import mime from 'mime';
import { getS3Client, PODCAST_BUCKET, S3_PODCAST_PREFIX } from './utils.js';

export const handler = async (event) => {
  console.log(`Event:\n${JSON.stringify(event)}`);

  const bucket = PODCAST_BUCKET;
  const key = getPodcastUploadKey(event);
  const s3Client = getS3Client();

  const mediaUrl = event?.episode?.media_url;
  console.info(`Downloading podcast from: ${mediaUrl}`);
  const incomeMessage = await downloadFromUrl(mediaUrl);
  const contentType = resolveContentType(event, mediaUrl, incomeMessage);
  console.log(
    `Uploading podcast to S3 URI: s3://${bucket}/${key} with Content-Type: ${contentType}`
  );
  throw new Error("Need to implement s3 upload");

  return updateEventState(event, key, contentType);
};

const getPodcastUploadKey = (event) => {
  return `${S3_PODCAST_PREFIX}${event?.episode?.id}/raw-media`;
};

const downloadFromUrl = (url) =>
  new Promise((resolve, reject) => {
    const req = request(url, (res) => {
      resolve(res);
    });
    req.on('error', (err) => {
      reject(err);
    });
    req.end();
  });

const resolveContentType = (event, mediaUrl, response) => {
  let contentType = event?.episode?.media_content_type;
  if (contentType === '' || contentType === 'application/octet-stream') {
    contentType = response.headers['Content-Type'];
  }
  contentType = contentType ?? mime.getType(mediaUrl);

  if (!contentType) {
    throw new Error(`Could not resolve content type for podcast`);
  }
  return contentType;
};

const updateEventState = (event, key, contentType) => {
  return {
    ...event,
    episode: {
      ...event.episode,
      media_key: key,
      media_content_type: contentType,
    },
  };
};

File: lambda/python/upload_podcast.py
import os
import contextlib
import mimetypes
import urllib.parse

import urllib3

from utils import PODCAST_BUCKET, S3_PODCAST_PREFIX, get_s3_client


class UnresolvableContentTypeException(Exception):
    pass


def lambda_handler(event, context):
    s3_bucket = PODCAST_BUCKET
    s3_key = _get_s3_podcast_upload_key(event)
    s3_client = get_s3_client()
    print(f'Downloading podcast from: {event["episode"]["media_url"]}')
    with _downloaded_podcast_stream(event['episode']['media_url']) as stream:
        content_type = _resolve_content_type(event, stream)
        print(
            f'Uploading podcast to S3 URI: s3://{s3_bucket}/{s3_key} '
            f'with Content-Type: {content_type}'
        )
        raise NotImplementedError(
            'Need to call s3_client.upload_fileobj() method to upload to S3'
        )
    _update_event_state(event, s3_key, content_type)
    return event


def _get_s3_podcast_upload_key(event):
    path = urllib.parse.urlparse(event['episode']['media_url']).path
    return f'{S3_PODCAST_PREFIX}{event["episode"]["id"]}/raw-media'


@contextlib.contextmanager
def _downloaded_podcast_stream(podcast_uri):
    http = urllib3.PoolManager()
    resp = http.request(
        "GET", podcast_uri, preload_content=False,
    )
    yield resp
    resp.release_conn()


def _resolve_content_type(event, response):
    content_type = event['episode']['media_content_type']
    if content_type in ['', 'application/octet-stream']:
        content_type = response.headers.get('Content-Type', '')
    if not content_type:
        content_type = mimetypes.guess_type(response.url)
    if not content_type:
        raise UnresolvableContentTypeException(
            'Could not resolve content type for podcast'
        )
    return content_type


def _update_event_state(event, s3_key, content_type):
    event['episode']['media_key'] = s3_key
    event['episode']['media_content_type'] = content_type
